src/indexer/utils/bitcoin/index.ts

```diff
@@ -0,0 +1,255 @@
+SPDX-License-Identifier: LGPL-3.0-only
+*/
+import FeeRepository from "indexer/repository/fee"
+import { sha256 } from "ethers"
+import { BigNumber } from "@ethersproject/bignumber"
+import { ObjectId } from "mongodb"
+import { TransferStatus } from "@prisma/client"
+import { RPCClient } from "rpc-bitcoin"
+import { BitcoinTypeTransfer, Block, DecodedDeposit, DecodedExecution, Transaction } from "../../../indexer/services/bitcoinIndexer/bitcoinTypes"
+import { BitcoinResource, Domain } from "../../../indexer/config"
+import ExecutionRepository from "../../repository/execution"
+import TransferRepository from "../../repository/transfer"
+import { logger } from "../../../utils/logger"
+import DepositRepository from "../../repository/deposit"
+import CoinMarketCapService from "../../../indexer/services/coinmarketcap/coinmarketcap.service"
+
+const WitnessV1Taproot = "witness_v1_taproot"
+const OP_RETURN = "nulldata"
+
+export async function saveEvents(
+  client: RPCClient,
+  block: Block,
+  domain: Domain,
+  executionRepository: ExecutionRepository,
+  transferRepository: TransferRepository,
+  depositRepository: DepositRepository,
+  feeRepository: FeeRepository,
+  coinMakerCapService: CoinMarketCapService,
+): Promise<void> {
+  for (const tx of block.tx) {
+    const depositData = decodeDepositEvent(tx, domain)
+    if (depositData) {
+      await saveDeposit(
+        tx.txid,
+        block.height,
+        block.mediantime,
+        domain.id,
+        depositData,
+        transferRepository,
+        depositRepository,
+        feeRepository,
+        coinMakerCapService,
+      )
+      continue
+    }
+    const executionData = await decodeExecutionEvent(tx, domain.resources as BitcoinResource[], client)
+    if (executionData) {
+      await saveProposalExecution(tx.txid, block.height, block.mediantime, domain.id, executionData, transferRepository, executionRepository)
+    }
+  }
+}
+
+function decodeDepositEvent(tx: Transaction, domain: Domain): DecodedDeposit | undefined {
+  let data = ""
+  let amount = BigInt(0)
+  let feeAmount = BigInt(0)
+  let resource
+
+  for (const vout of tx.vout) {
+    // Read OP_RETURN data
+    if (vout.scriptPubKey.type == OP_RETURN) {
+      // Extract OP_RETURN data (excluding OP_RETURN prefix)
+      const opReturnData = Buffer.from(vout.scriptPubKey.hex, "hex")
+      data = opReturnData.subarray(2).toString()
+      continue
+    }
+
+    const resourceFound = domain.resources.find(resource => resource.address === vout.scriptPubKey.address) as BitcoinResource
+    if (resourceFound) {
+      resource = resourceFound
+      if (vout.scriptPubKey.type == WitnessV1Taproot) {
+        amount = amount + BigInt(vout.value * 1e8)
+      }
+      continue
+    }
+
+    if (vout.scriptPubKey.address == domain.feeAddress) {
+      feeAmount = feeAmount + BigInt(vout.value * 1e8)
+    }
+  }
+
+  if (!resource || resource.feeAmount == undefined || feeAmount != BigInt(resource.feeAmount)) {
```

Comment:
Do these checks when checking resource and fee and handle accordingly instead of writing a complicated if for all cases.

Reply:
Not sure can I do that? Having multiple `vout`s in one tx, the first one might not be the relayer address leading to code not recognizing that it's a Sygma deposit.